### ISSUE TITLE
Minor edge case in Cminus leads to nans in output

### ISSUE THREAD

**Issue by ykwang1** [astroML/astroML#234]
Encountered a situation where cuml_x and cuml_y calculated by astroML.lumfunc.Cminus returns nans. This might only happen with bad luck.

Code to reproduce:
```
from astroML.lumfunc import Cminus
x, y, xmax, ymax = [10.02, 10.00], [14.97, 14.99], [10.03, 10.01], [14.98, 15.00]
Cminus(x, y, xmax, ymax)
```

Output:
`(array([0., 0.]), array([0., 0.]), array([ 0., nan]), array([ 0., nan]))`

Astrophysical context for the above input:
I would like to use the CMinus method to recover the true distributions of distance modulus (DM) and absolute magnitude (Mr), since I am limited by a flux limit of r < 25. Letting x<-DM and y<-Mr, I can calculate xmax and ymax (xmax = r_lim - y and vice versa). 

Since this example is at a peculiar case where the` x[0] > xmax[1]` this means `Nx[1] = np.sum(x[:1] < xmax[1])` to be 0, which causes `cuml_x = np.cumprod(1. + 1. / Nx)` evaluate to nan (see [source](https://github.com/astroML/astroML/blob/dbc4311fcd144cbd62dd1e32803e2b7738627f6b/astroML/lumfunc.py#L74)).

I think this only happens when, after [sorting ](https://github.com/astroML/astroML/blob/dbc4311fcd144cbd62dd1e32803e2b7738627f6b/astroML/lumfunc.py#L67), this unique case comes up where `np.sum(x[:j] < xmax[j]) == 0`, usually for very small j. Saw this while using simulated LSST data. I'm not familiar enough with Cminus to come up with a fix for this edge case, and seemed to show up most often when using astroML.lumfuc.bootstrap_Cminus.

**Comment by ykwang1** [comment]
Attached is a more substantive example. One possible way to work around this would be to discard the bootstrap samples in bootstrap_Cminus that result in distributions with nan. 
[break_cminus.csv.gz](https://github.com/astroML/astroML/files/6041496/break_cminus.csv.gz)

```
import pandas as pd
from astroML.lumfunc import Cminus

test_data = pd.read_csv('break_cminus.csv.gz')
x = test_data['xH']
y = test_data['yH']
xmax = test_data['xmaxH']
ymax = test_data['ymaxH']

xH, yH, cuml_xH, cuml_yH = Cminus(x, y, xmax, ymax)
print(cuml_yH)
```

**Comment by ivezic** [comment]
This case is fascinating! The Cminus method is 50 years old and as far as I know nobody ever reported this edge case. I agree with your assessment that the problem is due to the first point having a larger y than y_max for the second point. 
An easy and approximately correct fix would be to set Nx (and Ny) to 0.5 whenever they happen to be 0.

**Comment by bsipocz** [comment]
I don't yet have a full grasp of Cminus yet, so my proposal may be totally wrong. But what about setting the zeros to `np.inf`, similarly to what we already do with the first element (practically changing the array element to 1 instead of 3 for the cumulative product)? 
As we do a renormalization, it would change the result in prior to the last modified value:

```
In [6]: cuml_xH[:10]
Out[6]: 
array([0.00188867, 0.00377734, 0.00566602, 0.00755469, 0.00944336,
       0.01133203, 0.0132207 , 0.01510937, 0.01699805, 0.01888672])

In [7]: cuml_yH[:10]
Out[7]: 
array([0.00188867, 0.00377734, 0.00566602, 0.00755469, 0.00755469,
       0.01510937, 0.02266406, 0.03021875, 0.04532812, 0.0604375 ])

In [8]: xH[:10]
Out[8]: array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.])

In [9]: yH[:10]
Out[9]: array([ 0.,  1.,  2.,  3., inf,  1.,  2.,  3.,  2.,  3.])
```

vs

```
In [8]: 

In [8]: cuml_xH[:10]
Out[8]: 
array([0.00062956, 0.00125911, 0.00188867, 0.00251823, 0.00314779,
       0.00377734, 0.0044069 , 0.00503646, 0.00566602, 0.00629557])

In [9]: cuml_yH[:10]
Out[9]: 
array([0.00062956, 0.00125911, 0.00188867, 0.00251823, 0.00755469,
       0.01510937, 0.02266406, 0.03021875, 0.04532812, 0.0604375 ])

In [10]: xH[:10]
Out[10]: array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.])

In [11]: yH[:10]
Out[11]: array([0. , 1. , 2. , 3. , 0.5, 1. , 2. , 3. , 2. , 3. ])
```

**Comment by ivezic** [comment]
This might be a better way to fix it indeed. But if possible, let's try both and test them with David's break_cminus.csv data.

**Comment by bsipocz** [comment]
The above snippets are using David's data, the only difference is prior the last element where `np.sump(x[:j] < xmax[j]) == 0`.  In this datafile it's the 5th element out of the 6k+, but I think it would be possible to create a datafile where it happens later. Whether that datafile has realistic is another question.

**Comment by bsipocz** [comment]
PR with the `np.inf` fix is in #237, while the 0.5 approach is on the branch here if either of you prefers to check it out, too: https://github.com/bsipocz/astroML/tree/cminus_fix_nans_0.5
